File: config_watcher.py
#!/usr/bin/env python3
"""
Configuration Watcher for Atomic Clock Display

Watches for file changes and triggers live reloading of configuration.
"""

import os
import time
import threading
from typing import Callable, Optional
from pathlib import Path


class ConfigWatcher:
    """Watch configuration files for changes and trigger callbacks."""
    
    def __init__(self, config_path: str, callback: Callable[[], None], poll_interval: float = 1.0):
        """
        Initialize configuration watcher.
        
        Args:
            config_path: Path to configuration file to watch
            callback: Function to call when configuration changes
            poll_interval: Polling interval in seconds
        """
        self.config_path = Path(config_path)
        self.callback = callback
        self.poll_interval = poll_interval
        self.running = False
        self.thread = None
        self.last_mtime = None
        
        # Get initial modification time
        self._update_mtime()
    
    def _update_mtime(self):
        """Update the last modification time."""
        try:
            if self.config_path.exists():
                self.last_mtime = self.config_path.stat().st_mtime
            else:
                self.last_mtime = None
        except OSError:
            self.last_mtime = None
    
    def _watch_loop(self):
        """Main watching loop."""
        while self.running:
            try:
                if self.config_path.exists():
                    current_mtime = self.config_path.stat().st_mtime
                    
                    # Check if file was modified
                    if self.last_mtime is not None and current_mtime > self.last_mtime:
                        # Wait a brief moment to ensure file write is complete
                        time.sleep(0.1)
                        
                        # Verify file is still newer (avoid false triggers)
                        new_mtime = self.config_path.stat().st_mtime
                        if new_mtime > self.last_mtime:
                            self.last_mtime = new_mtime
                            try:
                                self.callback()
                            except Exception as e:
                                print(f"Error in config reload callback: {e}", file=sys.stderr)
                    else:
                        self.last_mtime = current_mtime
                else:
                    # File doesn't exist, reset mtime
                    self.last_mtime = None
                
                time.sleep(self.poll_interval)
                
            except Exception as e:
                print(f"Error watching config file: {e}", file=sys.stderr)
                time.sleep(self.poll_interval)
    
    def start(self):
        """Start watching for configuration changes."""
        if self.running:
            return
        
        self.running = True
        self.thread = threading.Thread(target=self._watch_loop, daemon=True)
        self.thread.start()
    
    def stop(self):
        """Stop watching for configuration changes."""
        self.running = False
        if self.thread:
            self.thread.join(timeout=2.0)
    
    def is_running(self) -> bool:
        """Check if watcher is currently running."""
        return self.running


class MultiConfigWatcher:
    """Watch multiple configuration files simultaneously."""
    
    def __init__(self, poll_interval: float = 1.0):
        """
        Initialize multi-configuration watcher.
        
        Args:
            poll_interval: Polling interval in seconds
        """
        self.poll_interval = poll_interval
        self.watchers = []
        self.running = False
        self.thread = None
    
    def add_config(self, config_path: str, callback: Callable[[], None]):
        """
        Add a configuration file to watch.
        
        Args:
            config_path: Path to configuration file
            callback: Function to call when configuration changes
        """
        watcher = ConfigWatcher(config_path, callback, self.poll_interval)
        self.watchers.append(watcher)
        
        # Start immediately if already running
        if self.running:
            watcher.start()
    
    def start(self):
        """Start watching all configuration files."""
        if self.running:
            return
        
        self.running = True
        for watcher in self.watchers:
            watcher.start()
    
    def stop(self):
        """Stop watching all configuration files."""
        self.running = False
        for watcher in self.watchers:
            watcher.stop()
    
    def is_running(self) -> bool:
        """Check if any watchers are currently running."""
        return self.running


if __name__ == '__main__':
    # Test configuration watcher
    import sys
    
    def test_callback():
        print("Configuration file changed!")
    
    if len(sys.argv) > 1:
        config_file = sys.argv[1]
        watcher = ConfigWatcher(config_file, test_callback)
        watcher.start()
        
        print(f"Watching {config_file} for changes...")
        print("Press Ctrl+C to stop")
        
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            print("\nStopping watcher...")
            watcher.stop()
    else:
        print("Usage: python config_watcher.py <config_file>")

File: config_watcher.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/inotify.h>
#include <errno.h>
#include <time.h>
#include <pthread.h>

#define EVENT_SIZE (sizeof(struct inotify_event))
#define BUF_LEN (1024 * (EVENT_SIZE + 16))

typedef struct {
    char *config_path;
    void (*callback)(void);
    int wd;  // inotify watch descriptor
    int fd;  // inotify file descriptor
    int running;
    pthread_t thread;
} ConfigWatcher;

typedef struct {
    ConfigWatcher *watchers;
    int count;
    int capacity;
    int running;
    pthread_t thread;
} MultiConfigWatcher;

static void *watch_loop(void *arg) {
    ConfigWatcher *watcher = (ConfigWatcher *)arg;
    char buffer[BUF_LEN];
    int length, i = 0;
    
    while (watcher->running) {
        length = read(watcher->fd, buffer, BUF_LEN);
        
        if (length < 0) {
            if (errno == EINTR) continue;
            perror("read");
            break;
        }
        
        while (i < length) {
            struct inotify_event *event = (struct inotify_event *)&buffer[i];
            
            if (event->mask & IN_MODIFY) {
                // Small delay to ensure file write is complete
                usleep(100000);  // 100ms
                
                if (watcher->callback) {
                    watcher->callback();
                }
            }
            
            i += EVENT_SIZE + event->len;
        }
        
        i = 0;
    }
    
    return NULL;
}

ConfigWatcher *config_watcher_create(const char *config_path, void (*callback)(void)) {
    ConfigWatcher *watcher = malloc(sizeof(ConfigWatcher));
    if (!watcher) return NULL;
    
    watcher->config_path = strdup(config_path);
    watcher->callback = callback;
    watcher->running = 0;
    
    // Initialize inotify
    watcher->fd = inotify_init();
    if (watcher->fd < 0) {
        perror("inotify_init");
        free(watcher->config_path);
        free(watcher);
        return NULL;
    }
    
    // Add watch
    watcher->wd = inotify_add_watch(watcher->fd, config_path, IN_MODIFY);
    if (watcher->wd < 0) {
        perror("inotify_add_watch");
        close(watcher->fd);
        free(watcher->config_path);
        free(watcher);
        return NULL;
    }
    
    return watcher;
}

void config_watcher_start(ConfigWatcher *watcher) {
    if (!watcher || watcher->running) return;
    
    watcher->running = 1;
    if (pthread_create(&watcher->thread, NULL, watch_loop, watcher) != 0) {
        perror("pthread_create");
        watcher->running = 0;
    }
}

void config_watcher_stop(ConfigWatcher *watcher) {
    if (!watcher || !watcher->running) return;
    
    watcher->running = 0;
    pthread_join(watcher->thread, NULL);
}

void config_watcher_destroy(ConfigWatcher *watcher) {
    if (!watcher) return;
    
    config_watcher_stop(watcher);
    
    if (watcher->wd >= 0) {
        inotify_rm_watch(watcher->fd, watcher->wd);
    }
    
    if (watcher->fd >= 0) {
        close(watcher->fd);
    }
    
    free(watcher->config_path);
    free(watcher);
}

// Test implementation
void test_callback(void) {
    printf("Configuration file changed!\n");
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <config_file>\n", argv[0]);
        return 1;
    }
    
    ConfigWatcher *watcher = config_watcher_create(argv[1], test_callback);
    if (!watcher) {
        fprintf(stderr, "Failed to create config watcher\n");
        return 1;
    }
    
    printf("Watching %s for changes...\n", argv[1]);
    printf("Press Ctrl+C to stop\n");
    
    config_watcher_start(watcher);
    
    // Wait for interrupt
    while (1) {
        sleep(1);
    }
    
    config_watcher_destroy(watcher);
    return 0;
}
